crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,265 @@
+// --- source
+// authors = ["Charlie Marsh"]
+// license = "MIT"
+// origin = "https://github.com/astral-sh/ruff/blob/03fb2e5ac1481e498f84474800b42a966e9843e1/crates/ruff_server/src/trace.rs"
+// ---
+
+//! The logging system for `air lsp`.
+//!
+//! Logs are controlled by the `air.logLevel` setting in VS Code,
+//! passed through `InitializeParams` in the arbitrary `initializationOptions` field.
+//!
+//! Logs are also controlled by the `AIR_LOG` environment variable. This is preferred
+//! over the extension setting, but it is unlikely to see them both be used together.
+//! This can be used by other LSP clients if there is no easy way to set `initializationOptions`.
+//!
+//! By default, we fallback to `"info"`.
+//!
+//! For VS Code and Zed, which are known to support `window/logMessage` well, logging will
+//! emit a `window/logMessage` message. Otherwise, logging will write to `stderr`,
+//! which should appear in the logs for most LSP clients.
+use core::str;
+use serde::Deserialize;
+use std::fmt::Display;
+use std::io::{Error as IoError, ErrorKind, Write};
+use tokio::sync::mpsc::unbounded_channel;
+use tower_lsp::lsp_types::ClientInfo;
+use tower_lsp::lsp_types::MessageType;
+use tower_lsp::Client;
+use tracing::level_filters::LevelFilter;
+use tracing_subscriber::fmt::time::LocalTime;
+use tracing_subscriber::{
+    fmt::{writer::BoxMakeWriter, MakeWriter},
+    layer::SubscriberExt,
+    Layer,
+};
+
+// TODO:
+// - Add `air.logLevel` as a VS Code extension option that sets the log level,
+//   and pass it through the arbitrary `initializationOptions` field of `InitializeParams`.
+
+const LOG_ENV_KEY: &str = "AIR_LOG";
+
+pub(crate) struct LogMessage {
+    contents: String,
+}
+
+pub(crate) type LogMessageSender = tokio::sync::mpsc::UnboundedSender<LogMessage>;
+pub(crate) type LogMessageReceiver = tokio::sync::mpsc::UnboundedReceiver<LogMessage>;
+
+pub(crate) struct LogState {
+    client: Client,
+    log_rx: LogMessageReceiver,
+}
+
+// Needed for spawning the loop
+unsafe impl Sync for LogState {}
+
+impl LogState {
+    pub(crate) fn new(client: Client) -> (Self, LogMessageSender) {
+        let (log_tx, log_rx) = unbounded_channel::<LogMessage>();
+        let state = Self { client, log_rx };
+        (state, log_tx)
+    }
+
+    /// Start the log loop
+    ///
+    /// Takes ownership of log state and start the low-latency log loop.
+    ///
+    /// We use `MessageType::LOG` to prevent the middleware from adding its own
+    /// timestamp and log level labels. We add that ourselves through tracing.
+    pub(crate) async fn start(mut self) {
+        while let Some(message) = self.log_rx.recv().await {
+            self.client
+                .log_message(MessageType::LOG, message.contents)
```

Comment:
We always forward log messages of type `LOG` to the middleware.

Sending `LOG` messages means the middleware calls the raw `outputChannel.appendLine()` here:
https://github.com/microsoft/vscode-languageserver-node/blob/906f5fb306e1f6059cbdcb1efd962647222b5867/client/src/common/client.ts#L1281-L1297

I prefer that over going through `ERROR` or `WARN` in the middleware, which goes through this `logOutputMessage()` helper:
https://github.com/microsoft/vscode-languageserver-node/blob/906f5fb306e1f6059cbdcb1efd962647222b5867/client/src/common/client.ts#L1179-L1187

That helper unconditionally prepends the log level and a timestamp to the message, but we already do this.

It will also show a notification as a toast message to the user if `showNotification` is `true` or `"force"`, but it's always `false` when processing server log messages. This means that literally the only thing we would get from the middleware by sending `MessageType::ERROR` is for it to automatically prepend `[Error - <time>]` to our messages, so it seems fine to skip this.